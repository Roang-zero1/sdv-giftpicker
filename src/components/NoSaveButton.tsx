import * as charactersActions from '../actions/charactersActions';
import * as itemsActions from '../actions/itemsActions';
import * as statusActions from '../actions/statusActions';

import * as React from 'react';
import { Component } from 'react';

import { connect, Dispatch } from 'react-redux';
import { Button } from 'reactstrap';
import { bindActionCreators } from 'redux';
import { IGiftTastes } from '../common/types';

/* tslint:disable-next-line:no-var-requires */
const giftIDs: number[] = require('../data/Gifts.json');
/* tslint:disable-next-line:no-var-requires */
const giftTastes: IGiftTastes = require('../data/GiftTastes.json');

export interface IDispatchProps {
  charactersActions: typeof charactersActions;
  itemsActions: typeof itemsActions;
  statusActions: typeof statusActions;
}

export interface IProps extends IDispatchProps {
  inline?: boolean;
  text?: string;
  className?: string;
}

export class NoSaveButton extends Component<IProps> {
  constructor(props: IProps) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }
  public render() {
    const { className, text, inline } = this.props;
    return (
      <Button
        color="secondary"
        onClick={this.handleClick}
        className={className}
        size={inline ? 'sm' : undefined}
      >
        {text ? text : 'Use without save'}
      </Button>
    );
  }

  public handleClick() {
    this.props.statusActions.setLoading(true);
    this.props.statusActions.setSaveGame(false);
    for (const char of Object.keys(giftTastes)) {
      this.props.charactersActions.setGiftCount({ char, count: 0 });
    }
    const items = {};
    for (const gift of giftIDs) {
      items[gift] = 999;
    }
    this.props.itemsActions.updateItems(items);

    this.props.statusActions.setLoading(false);
    this.props.statusActions.setIntroChosen(true);
  }
}

function mapDispatchToProps(dispatch: Dispatch): IDispatchProps {
  return {
    charactersActions: bindActionCreators(charactersActions, dispatch),
    itemsActions: bindActionCreators(itemsActions, dispatch),
    statusActions: bindActionCreators(statusActions, dispatch)
  };
}

export default connect(
  null,
  mapDispatchToProps
)(NoSaveButton);
