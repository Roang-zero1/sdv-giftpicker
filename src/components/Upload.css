#upload {
  display: none;
}

.inline {
  display: inline-block;
}
