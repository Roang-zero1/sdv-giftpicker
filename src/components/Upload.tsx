import * as charactersActions from '../actions/charactersActions';
import * as itemsActions from '../actions/itemsActions';
import * as statusActions from '../actions/statusActions';

import * as React from 'react';
import { Component, RefObject } from 'react';

import * as $ from 'jquery';
import { connect, Dispatch } from 'react-redux';
import { Button } from 'reactstrap';
import { bindActionCreators } from 'redux';
import styled from 'styled-components';
import { IGiftTastes, RootState } from '../common/types';

/* tslint:disable-next-line:no-var-requires */
const giftIDs: number[] = require('../data/Gifts.json');
/* tslint:disable-next-line:no-var-requires */
const giftTastes: IGiftTastes = require('../data/GiftTastes.json');

interface IWindow extends Window {
  DOMParser: DOMParser | undefined;
  ActiveXObject: any;
}

export interface IDispatchProps {
  charactersActions: typeof charactersActions;
  itemsActions: typeof itemsActions;
  statusActions: typeof statusActions;
}

export interface IProps extends IDispatchProps {
  inline?: boolean;
  text?: string;
  className: string;
}

interface IWrapperProps {
  inline?: boolean;
}

const Wrapper = styled<IWrapperProps, 'div'>('div')`
  display: ${props => (props.inline ? 'inline-block' : undefined)};
`;

const HiddenInput = styled.input`
  display: none;
`;

class Upload extends Component<IProps> {
  private inputRef: RefObject<HTMLInputElement>;
  // TODO: Properly type this
  private items: any;
  constructor(props: any) {
    super(props);
    this.inputRef = React.createRef();
  }

  public render() {
    const { inline, className, text } = this.props;
    return (
      <Wrapper inline={inline} className={className}>
        <HiddenInput
          type="file"
          onChange={this.handleUpload}
          innerRef={this.inputRef}
        />
        <Button
          color="primary"
          size={inline ? 'sm' : 'lg'}
          onClick={this.handleClick}
          className={className}
        >
          {text ? text : 'Upload save-game'}
        </Button>
      </Wrapper>
    );
  }

  private parseItems(xmlDoc: XMLDocument, searchString: string) {
    const items = this.items;
    $(xmlDoc)
      .find(searchString)
      .each(function() {
        let id = 0;
        try {
          id = parseInt(
            $(this)
              .find('parentSheetIndex')
              .text(),
            10
          );
          if (giftIDs.indexOf(id) > -1) {
            const count = parseInt(
              $(this)
                .find('Stack')
                .text(),
              10
            );
            items[id] = (items[id] || 0) + count;
          }
        } catch (err) {
          // TODO: Add proper error handling
          console.log('Failed to get item count for ' + id);
        }
      });
  }

  private gatherItems(xmlDoc: XMLDocument) {
    this.items = {};
    this.parseItems(xmlDoc, 'player > items > Item[xsi\\:type="Object"]');
    this.parseItems(
      xmlDoc,
      'locations > GameLocation[xsi\\:type="FarmHouse"] > fridge > items > Item[xsi\\:type="Object"]'
    );
    this.parseItems(
      xmlDoc,
      'Object[xsi\\:type="Chest"] > items > Item[xsi\\:type="Object"]'
    );
    this.props.itemsActions.updateItems(this.items);
    delete this.items;
  }

  private findGiftCount = (oDOM: Document) => {
    const nodes = oDOM.evaluate(
      'SaveGame/player/friendships/item',
      oDOM,
      null,
      XPathResult.ANY_TYPE,
      null
    );
    let result = nodes.iterateNext();
    while (result) {
      const char = oDOM.evaluate(
        'key/string',
        result,
        null,
        XPathResult.STRING_TYPE,
        null
      ).stringValue;

      const count = oDOM.evaluate(
        'value/ArrayOfInt/int[2]',
        result,
        null,
        XPathResult.NUMBER_TYPE,
        null
      ).numberValue;

      if (char && char in giftTastes && Number.isInteger(count)) {
        this.props.charactersActions.setGiftCount({
          char,
          count
        });
      }

      result = nodes.iterateNext();
    }
  };

  private handleUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files) {
      const file = event.target.files[0];
      const reader = new FileReader();

      reader.onloadstart = this.onLoadStart;
      reader.onprogress = this.onProgress;
      reader.onload = this.onLoad;

      reader.readAsText(file);
    }
  };

  private handleClick = () => {
    if (this.inputRef.current) {
      this.inputRef.current.click();
    }
  };

  private onLoadStart = (ev: FileReaderProgressEvent) => {
    this.props.statusActions.setLoading(true);
  };

  private onProgress = (ev: FileReaderProgressEvent) => {
    if (ev.lengthComputable) {
      this.props.statusActions.setLoading(true);
    }
  };

  private parseXml = (xmlString: string): Document => {
    let oDOM;
    const win = window as IWindow;
    if (typeof win.DOMParser !== 'undefined') {
      const parser = new DOMParser();
      return parser.parseFromString(xmlString, 'text/xml');
    } else if (
      typeof win.ActiveXObject !== 'undefined' &&
      new win.ActiveXObject('Microsoft.XMLDOM')
    ) {
      oDOM = new win.ActiveXObject('Microsoft.XMLDOM');
      oDOM.async = false;
      oDOM.loadXML(xmlString);
      return oDOM;
    } else {
      throw new Error('No XML parser found');
    }

    return oDOM;
  };

  private onLoad = (ev: FileReaderProgressEvent) => {
    this.props.statusActions.setLoading(true);
    try {
      if (ev.target) {
        const xmlDoc = $.parseXML(ev.target.result);
        const oDOM = this.parseXml(ev.target.result);

        console.log(
          oDOM.documentElement.nodeName === 'parsererror'
            ? 'error while parsing'
            : oDOM.documentElement.nodeName
        );

        this.gatherItems.call(this, xmlDoc);
        this.props.statusActions.setLoading(true);
        this.findGiftCount(oDOM);
        this.props.statusActions.setLoading(true);
        this.props.statusActions.setLoading(false);
        this.props.statusActions.setSaveGame(true);
        this.props.statusActions.setIntroChosen(true);
        if (this.inputRef.current) {
          this.inputRef.current.value = '';
        }
      }
    } catch (err) {
      // TODO: Add proper error handling
      console.log('Failed to parse file');
    }
  };
}

function mapStateToProps(state: RootState) {
  return {};
}

function mapDispatchToProps(dispatch: Dispatch) {
  return {
    charactersActions: bindActionCreators(charactersActions, dispatch),
    itemsActions: bindActionCreators(itemsActions, dispatch),
    statusActions: bindActionCreators(statusActions, dispatch)
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Upload);
