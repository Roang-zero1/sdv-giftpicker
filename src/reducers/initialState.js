export default {
  characters: {},
  status: { intro: false, save: false }
};
