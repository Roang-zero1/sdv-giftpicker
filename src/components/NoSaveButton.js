import './Upload.css';

import * as charactersActions from '../actions/charactersActions';
import * as itemsActions from '../actions/itemActions';
import * as statusActions from '../actions/statusActions';

import React, { Component } from 'react';

import { Button } from 'reactstrap';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import giftIDs from '../data/Gifts';
import giftTastes from '../data/GiftTastes';

class NoSaveButton extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }
  render(props) {
    return (
      <Button
        color="secondary"
        onClick={this.handleClick}
        className={this.props.className}
        size={this.props.inline ? 'sm' : null}
      >
        {this.props.text ? this.props.text : 'Use without save'}
      </Button>
    );
  }

  handleClick(event) {
    this.props.statusActions.setLoading(true);
    this.props.statusActions.setSaveGame(false);
    for (let char in giftTastes) {
      this.props.charactersActions.setGiftCount(char, 0);
    }
    let items = {};
    for (let gift of giftIDs) {
      items[gift] = 999;
    }
    this.props.itemsActions.updateItems(items);

    this.props.statusActions.setLoading(false);
    this.props.statusActions.setIntroChosen(true);
  }
}

function mapStateToProps(state) {
  return {};
}

function mapDispatchToProps(dispatch) {
  return {
    itemsActions: bindActionCreators(itemsActions, dispatch),
    statusActions: bindActionCreators(statusActions, dispatch),
    charactersActions: bindActionCreators(charactersActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(NoSaveButton);
