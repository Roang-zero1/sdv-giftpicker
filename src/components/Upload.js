import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { Button } from 'reactstrap';
import $ from 'jquery';

import giftIDs from '../data/Gifts.js';

import * as itemsActions from '../actions/itemActions';
import * as statusActions from '../actions/statusActions';
import * as charactersActions from '../actions/charactersActions';

import './Upload.css';

class Upload extends Component {
  constructor(props) {
    super(props);
    this.handleUpload = this.handleUpload.bind(this);
  }
  render(props) {
    return (
      <div className={classNames({ inline: this.props.inline })}>
        <input
          type="file"
          id="upload"
          onChange={this.handleUpload}
          ref={upload => {
            this.upload = upload;
          }}
        />
        <Button
          color="primary"
          size={this.props.inline ? 'sm' : 'lg'}
          onClick={() => {
            this.upload.click();
          }}
          disabled={this.props.disabled}
          className={this.props.className}
        >
          {this.props.text ? this.props.text : 'Upload save-game'}
        </Button>
      </div>
    );
  }

  parseItems(xmlDoc, searchString) {
    const items = this.items;
    $(xmlDoc)
      .find(searchString)
      .each(function() {
        try {
          var id = parseInt(
            $(this)
              .find('parentSheetIndex')
              .text(),
            10
          );
          if (giftIDs.includes(id)) {
            var count = parseInt(
              $(this)
                .find('Stack')
                .text(),
              10
            );
            items[id] = (items[id] || 0) + count;
          }
        } catch (err) {
          // TODO: Add proper error handling
          console.log('Failed to get item count for ' + id);
        }
      });
  }

  gatherItems(xmlDoc) {
    this.items = {};
    this.parseItems(xmlDoc, 'player > items > Item[xsi\\:type="Object"]');
    this.parseItems(
      xmlDoc,
      'locations > GameLocation[xsi\\:type="FarmHouse"] > fridge > items > Item[xsi\\:type="Object"]'
    );
    this.parseItems(
      xmlDoc,
      'Object[xsi\\:type="Chest"] > items > Item[xsi\\:type="Object"]'
    );
    this.props.itemsActions.updateItems(this.items);
    delete this.items;
  }

  findGiftCount(xmlDoc) {
    const setGiftCount = this.props.charactersActions.setGiftCount;
    $(xmlDoc)
      .find('player > friendships > item')
      .each(function() {
        try {
          var who = $(this)
            .find('key > string')
            .html();
          var num = parseInt(
            $(this)
              .find('value > ArrayOfInt > int')
              .first()
              .next()
              .text(),
            10
          );
          setGiftCount(who, num);
        } catch (err) {
          // TODO: Add proper error handling
          console.log('Failed to update data for ' + who + '\n' + err);
        }
      });
  }

  handleUpload(event) {
    var target = event.target;
    var file = event.target.files[0];
    var reader = new FileReader();

    const instance = this;

    reader.onloadstart = function(e) {
      instance.props.statusActions.setLoading(true);
    };

    reader.onprogress = function(e) {
      if (e.lengthComputable) {
        instance.props.statusActions.setLoading(true);
      }
    };

    reader.onload = function(e) {
      instance.props.statusActions.setLoading(true);
      try {
        var xmlDoc = $.parseXML(e.target.result);
        instance.gatherItems.call(instance, xmlDoc);
        instance.props.statusActions.setLoading(true);
        instance.findGiftCount.call(instance, xmlDoc);
        instance.props.statusActions.setLoading(true);
        instance.props.statusActions.setLoading(false);
        instance.props.statusActions.setSaveGame(true);
        instance.props.statusActions.setIntroChosen(true);
        target.value = '';
      } catch (err) {
        // TODO: Add proper error handling
        console.log('Failed to parse file');
      }
    };

    reader.readAsText(file);
  }
}

function mapStateToProps(state) {
  return {};
}

function mapDispatchToProps(dispatch) {
  return {
    itemsActions: bindActionCreators(itemsActions, dispatch),
    statusActions: bindActionCreators(statusActions, dispatch),
    charactersActions: bindActionCreators(charactersActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Upload);
