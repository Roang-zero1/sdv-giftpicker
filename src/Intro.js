import React, { Component } from 'react';

import { Jumbotron, Container, Col } from 'reactstrap';

import Upload from './components/Upload';
import NoSaveButton from './components/NoSaveButton';
import About from './components/About';

import './Intro.css';

class Intro extends Component {
  render() {
    return (
      <main className="App">
        <Jumbotron className="header">
          <Container>
            <h1 className="display-3">SDV Gift Picker</h1>
            <p>
              Use this site to check which gifts you have available for the
              characters of{' '}
              <a href="http://stardewvalley.net/">Stardew Valley</a>. The save
              is no uploaded and processed locally.
            </p>
            <Upload className="mb-2" />
            <NoSaveButton />
          </Container>
        </Jumbotron>
        <Container>
          <Col xs="12">
            <h2>Help</h2>
            <p>
              Please use the full save file named with your farmer's name and a
              9-digit ID number (e.g.
              <code>Fred_148093307</code>); do not use the{' '}
              <code>SaveGameInfo</code> file as it does not contain all the
              necessary information.
            </p>
            <p>Default save file locations are:</p>
            <div>
              <ul>
                <li>
                  Windows: <code>%AppData%\StardewValley\Saves\</code>
                </li>
                <li>
                  Mac OSX &amp; Linux:{' '}
                  <code>~/.config/StardewValley/Saves/</code>
                </li>
              </ul>
            </div>
          </Col>
          <About />
        </Container>
      </main>
    );
  }
}
export default Intro;
