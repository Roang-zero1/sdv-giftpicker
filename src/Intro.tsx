import './Intro.css';

import * as React from 'react';
import { Component } from 'react';
import { Col, Container, Jumbotron } from 'reactstrap';

import About from './components/About';
import NoSaveButton from './components/NoSaveButton';
import Upload from './components/Upload';

class Intro extends Component {
  public render() {
    return (
      <main className="App">
        <Jumbotron className="header">
          <Container>
            <h1 className="display-3">SDV Gift Picker</h1>
            <p>
              Use this site to check which gifts you have available for the
              characters of{' '}
              <a href="http://stardewvalley.net/">Stardew Valley</a>. The save
              is no uploaded and processed locally.
            </p>
            <Upload className="mb-2" />
            <NoSaveButton />
          </Container>
        </Jumbotron>
        <Container>
          <Col xs="12">
            <h2>Help</h2>
            <p>
              Please use the full save file named with your farmer's name and a
              9-digit ID number (e.g.
              <code>Fred_148093307</code>); do not use the{' '}
              <code>SaveGameInfo</code> file as it does not contain all the
              necessary information.
            </p>
            <p>Default save file locations are:</p>
            <div>
              <ul>
                <li>
                  Windows: <code>%AppData%\StardewValley\Saves\</code>
                </li>
                <li>
                  Mac OSX &amp; Linux:{' '}
                  <code>~/.config/StardewValley/Saves/</code>
                </li>
              </ul>
            </div>
          </Col>
          <About />
        </Container>
      </main>
    );
  }
}
export default Intro;
