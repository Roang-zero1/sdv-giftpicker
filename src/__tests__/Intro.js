import { mount, shallow } from 'enzyme';

import Intro from '../Intro';
import NoSaveButton from '../components/NoSaveButton';
import { Provider } from 'react-redux';
import React from 'react';
import Upload from '../components/Upload';
import configureStore from 'redux-mock-store';
import initialState from '../reducers/initialState';
import renderer from 'react-test-renderer';

describe('Intro --- Shallow render component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Intro />);
  });

  it('should render the one component', () => {
    expect(wrapper.length).toEqual(1);
  });
});

describe('Intro --- Render component', () => {
  const mockStore = configureStore();
  let component, store;
  beforeEach(() => {
    store = mockStore(initialState);
    component = mount(
      <Provider store={store}>
        <Intro />
      </Provider>
    );
  });

  it('should render self and sub-components', () => {
    expect(component).toHaveLength(1);

    expect(component.find(Upload)).toHaveLength(1);
    expect(component.find(NoSaveButton)).toHaveLength(1);
  });

  it('should be the same as the last snapshot', () => {
    const renderValue = renderer.create(component).toJSON();

    expect(renderValue).toMatchSnapshot();
  });
});
