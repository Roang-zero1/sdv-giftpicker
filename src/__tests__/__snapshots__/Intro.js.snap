// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intro --- Render connected component should be the same as the last snapshot 1`] = `
<main
  className="App"
>
  <div
    className="header jumbotron"
  >
    <div
      className="container"
    >
      <h1
        className="display-3"
      >
        SDV Gift Picker
      </h1>
      <p>
        Use this site to check which gifts you have available for the characters of
         
        <a
          href="http://stardewvalley.net/"
        >
          Stardew Valley
        </a>
        . The save is no uploaded and processed locally.
      </p>
      <div
        className=""
      >
        <input
          id="upload"
          onChange={[Function]}
          type="file"
        />
        <button
          className="mb-2 btn btn-primary btn-lg"
          disabled={undefined}
          onClick={[Function]}
          type="button"
        >
          Upload save-game
        </button>
      </div>
      <button
        className="btn btn-secondary"
        onClick={[Function]}
        type="button"
      >
        Use without save
      </button>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="col-12"
    >
      <h2>
        Help
      </h2>
      <p>
        Please use the full save file named with your farmer's name and a 9-digit ID number (e.g.
        <code>
          Fred_148093307
        </code>
        ); do not use the
         
        <code>
          SaveGameInfo
        </code>
         file as it does not contain all the necessary information.
      </p>
      <p>
        Default save file locations are:
      </p>
      <div>
        <ul>
          <li>
            Windows: 
            <code>
              %AppData%\\StardewValley\\Saves\\
            </code>
          </li>
          <li>
            Mac OSX & Linux:
             
            <code>
              ~/.config/StardewValley/Saves/
            </code>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="about col-12"
    >
      <h2>
        About
      </h2>
      <p>
        SDV Gift Calculator by
         
        <a
          href="https://www.reddit.com/user/Roang_zero1/"
        >
          Roang_Zero1
        </a>
        .
      </p>
      <p>
        This app was inspired by 
        <strong>
          Stardew Checkup
        </strong>
         available at
         
        <a
          href="https://mouseypounds.github.io/stardew-checkup/"
        >
          https://mouseypounds.github.io/stardew-checkup/
        </a>
      </p>
      <p>
        Using icons by 
        <a
          href="https://fontawesome.com/l"
        >
          Font Awesome
        </a>
         
        licensed under
         
        <a
          href="https://creativecommons.org/licenses/by/4.0/"
        >
          CC BY 4.0
        </a>
      </p>
    </div>
  </div>
</main>
`;
